### PR TITLE
Implement Dataset and Corpus data loaders.

### LINKED ISSUE #416: Unable to parse meta.json files from  the mushroom and bikeshare datasets
Hi,

I downloaded all the data defined in this script https://github.com/DistrictDataLabs/yellowbrick/blob/develop/examples/download.py. Most of them have a meta.json file inside their folder but for two of them (mushroom and bikeshare datasets) I have a JSONDecodeError exception when I try to parse it with the json module of the standard library. 

### Code

This is an example code that can reproduce the error.
 ```python
import json

with open('data/bikeshare/meta.json') as file:
    meta_data = json.load(file)
meta_data
```

### PR DESCRIPTION
**Pull request by bbengfort**
This is a significant change that perhaps we should save for the 1.0 release of Yellowbrick. Datasets, instead of being simple numpy arrays, are now complex objects that provide a lot of functionality similar to the sklearn Bunch object; e.g. access to README and meta information, the ability to transform data into arrays or data frames and more.

This represents a complete overhaul of how datasets in Yellowbrick are managed and will require updates to the documentation and tests.

Still a work on progress. 

Related to #527 #512 #500 #441 #424 

Fixes #416

### CODE REVIEW

**Comment by rebeccabilbro** [comment]
@bbengfort - very excited about this one! When it's ready, this might be a good candidate for a live code review with the oz maintainers (e.g. with Slack or Hangouts) so that everyone can get synced up together & we can strategize on follow-up tasks.

**Comment by bbengfort** [comment]
Goals:

- [ ] add a `legacy` parameter to the `load_*` methods so that current examples aren't broken 
- [ ] add deprecation warnings for the datasets examples 
- [ ] save datasets as `npz` files for better matrix loading of data 
- [ ] use the meta.json file to produce `X` and `y` datasets 
- [ ] implement `to_matrix` and `to_pandas` methods of the dataset 
- [ ] better implementation of `Corpus` dataset 
- [ ] have tests use the new datasets code instead of `tests.datasets`

**Comment by Kautumn06** [comment]
@bbengfort One small fix for the `credit` dataset, is that when I run `data = load_credit()` the list `data['feature_names']` currently has one extra feature, `id`. If I run `len(data["feature_names"))` I get 25 but `data['data'].shape` returns `(30,000, 24)` and it contains `default` which is the target variable. 

Just to be sure, I did check our original `credit` dataset using `load_data("credit")` and the UCI Machine Learning Repository and `id` wasn't a feature in either of those. Let me know if you have any questions!

**Comment by rebeccabilbro** [comment]
Sincere apologies for sitting on this one so long! I will aim to get something on the calendar soon for the @DistrictDataLabs/team-oz-maintainers to take a look at this one together.

In the meantime, I've found another dataset I think will be nice for our clustering visualizers. It's a wrangled version of the [Walking Activity Dataset](https://archive.ics.uci.edu/ml/datasets/User+Identification+From+Walking+Activity) from the UCI Machine Learning Repository. This data captures the time-step, x acceleration, y acceleration, and z acceleration of people walking. In the original dataset, there's a separate file for each of the 22 participants. In [my version](https://github.com/rebeccabilbro/rebeccabilbro.github.io/blob/master/data/all_walkers.csv.zip), it's all wrangled into a single file, with an additional column that gives the `participant_id`. [Wrangling gist here](https://gist.github.com/rebeccabilbro/82cb2c03cebaad406ae3b0184e4b56bd), just in case we need to reproduce or if you want to wrangle it any differently from me.

**Comment by bbengfort** [comment]
@DistrictDataLabs/team-oz-maintainers this PR is ready to be reviewed (presuming the tests pass). Because this is such a major overhaul, it was recommended that we do a Google Hangouts to code review it rather than reviewing it asynchronously. 

I'm trying to get a lot of v1.0 stuff done because @rebeccabilbro and I are going to PyData Miami to present on Yellowbrick. Would it at all be possible to do the video review this week? I'm free:

- Tonight (Tuesday, January 1)
- Tomorrow Night
- Friday Night
- Saturday

**Comment by rebeccabilbro** [comment]
Amazing @bbengfort - thank you for your persistence with this and look forward to discussing! I could chat tonight, Friday eve, or Sat.

**Comment by Kautumn06** [comment]
I am so excited for this update @bbengfort! I'm happy to talk this evening, Friday, or Saturday.

File: tests/test_datasets/test_download.py [hunk 1]
```diff
@@ -1,44 +1,46 @@
...
+
+
+@pytest.mark.parametrize("loader", [
+        load_bikeshare, load_concrete, load_credit, load_energy, load_game,
+        load_mushroom, load_occupancy, load_spam, load_walking, load_hobbies,
+    ], ids=lambda l: l.__name__)
+def test_loader_download(tmpdir, loader):
     """
-    Test the dataset loading functions
+    Test download of dataset when it does not exist
```
Comment (bbengfort) [inline tests/test_datasets/test_download.py]: Mention will fail if no internet connection.

File: tests/test_datasets/test_loaders.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+    def test_load_concrete(self):
+        """
+        Test loading the concrete regression dataset
+        """
+        data = load_concrete(return_dataset=True)
+        assert_valid_dataset(data, "concrete")
+
+        X, y = load_concrete()
+        assert X.shape == (1030, 8)
```
Comment (bbengfort) [inline tests/test_datasets/test_loaders.py]: Will fail if dataset versions have changed unexpectedly, causing shape to change.

**Comment by bbengfort** [comment]
Oops and we should remember to add to the changelog!

**Pull request merged**